
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>genoml.nlp package &#8212; genoml 0.0.1 documentation</title>
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="genoml-nlp-package">
<h1>genoml.nlp package<a class="headerlink" href="#genoml-nlp-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-genoml.nlp.bedtools">
<span id="genoml-nlp-bedtools-module"></span><h2>genoml.nlp.bedtools module<a class="headerlink" href="#module-genoml.nlp.bedtools" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="genoml.nlp.bedtools.BedTool">
<em class="property">class </em><code class="descclassname">genoml.nlp.bedtools.</code><code class="descname">BedTool</code><span class="sig-paren">(</span><em>fn=None</em>, <em>from_string=False</em>, <em>remote=False</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.bedtools.BedTool" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">pybedtools.bedtool.BedTool</span></code></p>
<p>Class to wrap pybedtools.bt for use with machine learning and
natural language processing</p>
<dl class="attribute">
<dt id="genoml.nlp.bedtools.BedTool.alphabet">
<code class="descname">alphabet</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.bedtools.BedTool.alphabet" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="genoml.nlp.bedtools.BedTool.complement">
<code class="descname">complement</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.bedtools.BedTool.complement" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="genoml.nlp.bedtools.BedTool.create_corpus">
<code class="descname">create_corpus</code><span class="sig-paren">(</span><em>intervals</em>, <em>k</em>, <em>fasta=None</em>, <em>upstream=0</em>, <em>downstream=0</em>, <em>offset=1</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.bedtools.BedTool.create_corpus" title="Permalink to this definition">¶</a></dt>
<dd><p>Create corpus of kmer tokens from intervals
:param intervals (list): list of interval indexes or pybedtools Interval objects
:param k (integer): size of each kmer
:option fasta (string/io.TextIOWrapper): path to fasta file or stream
:option upstream (integer): number of bases upstream from each interval to read
:option downstream (integer): number of bases downstream from each interval to read
:option offset (integer): shift offset between kmers in each interval
:return corpus (list): corpus of kmer tokens from intervals</p>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.bedtools.BedTool.read_expression_level">
<code class="descname">read_expression_level</code><span class="sig-paren">(</span><em>interval</em>, <em>expr=None</em>, <em>sample_regex=None</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.bedtools.BedTool.read_expression_level" title="Permalink to this definition">¶</a></dt>
<dd><p>Read expression level for gene corresponding to a given interval
:param interval (int/Interval): index of interval or pybedtools Interval object
:option expr (string): path to normalized feature count table
:option sample_regex (string): regex to filter by sample name
:return mean_expr_level (float): mean expression level for gene corresponding to interval</p>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.bedtools.BedTool.read_kmers">
<code class="descname">read_kmers</code><span class="sig-paren">(</span><em>interval</em>, <em>k</em>, <em>fasta=None</em>, <em>upstream=0</em>, <em>downstream=0</em>, <em>offset=1</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.bedtools.BedTool.read_kmers" title="Permalink to this definition">¶</a></dt>
<dd><p>Read kmers from sequence from annotated interval in fasta file
:param interval (int/Interval): index of interval or pybedtools Interval object
:param k (integer): size of each kmer
:option fasta (string/io.TextIOWrapper): path to fasta file or stream
:option upstream (integer): number of bases upstream from interval to read
:option downstream (integer): number of bases downstream from interval to read
:option offset (integer): shift offset between kmers in interval
:return kmers (KmerList): list of kmers in interval</p>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.bedtools.BedTool.read_seq">
<code class="descname">read_seq</code><span class="sig-paren">(</span><em>interval</em>, <em>fasta=None</em>, <em>upstream=0</em>, <em>downstream=0</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.bedtools.BedTool.read_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Read sequence from annotated interval in fasta file
:param interval (int/Interval): index of interval or pybedtools Interval object
:option fasta (string/io.TextIOWrapper): path to fasta file or stream
:option upstream (integer): number of bases upstream from interval to read
:option downstream (integer): number of bases downstream from interval to read
:return sequence (string): raw interval sequence from fasta file</p>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.bedtools.BedTool.read_tokens">
<code class="descname">read_tokens</code><span class="sig-paren">(</span><em>interval</em>, <em>k</em>, <em>fasta=None</em>, <em>upstream=0</em>, <em>downstream=0</em>, <em>offset=1</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.bedtools.BedTool.read_tokens" title="Permalink to this definition">¶</a></dt>
<dd><p>Read tokens from sequence from annotated interval in Fasta file
:param interval (int/Interval): index of interval or pybedtools Interval object
:param k (integer): size of each kmer
:option fasta (string/io.TextIOWrapper): path to fasta file or stream
:option upstream (integer): number of bases upstream from interval to read
:option downstream (integer): number of bases downstream from interval to read
:option offset (integer): shift offset between kmers in interval
:return tokens (KmerList.TokenList): list of kmer tokens in interval</p>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.bedtools.BedTool.remove_invalid">
<code class="descname">remove_invalid</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.bedtools.BedTool.remove_invalid" title="Permalink to this definition">¶</a></dt>
<dd><p>Wrapper for bt.remove_invalid(<a href="#id1"><span class="problematic" id="id2">*</span></a>args, <a href="#id3"><span class="problematic" id="id4">**</span></a>kwargs)</p>
<p>Removes invalid entries from GTF/GFF files (i.e. chromosomes with
negative coordinates or features of zero length</p>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.bedtools.BedTool.saveas">
<code class="descname">saveas</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.bedtools.BedTool.saveas" title="Permalink to this definition">¶</a></dt>
<dd><p>Wrapper for bt.saveas(<a href="#id5"><span class="problematic" id="id6">*</span></a>args, <a href="#id7"><span class="problematic" id="id8">**</span></a>kwargs)
Creates tempfile to store intermediate data, allows IntervalIterator to reset</p>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.bedtools.BedTool.subset_feature">
<code class="descname">subset_feature</code><span class="sig-paren">(</span><em>feature_type</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.bedtools.BedTool.subset_feature" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract feature type from each gene
:param feature_type (string): type of gene feature (e.g. five_prime_utr, three_prime_utr, exon, etc.)
:return feature_iter (bt): pybedtools interval iterator limited to region_type intervals</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-genoml.nlp.kmers">
<span id="genoml-nlp-kmers-module"></span><h2>genoml.nlp.kmers module<a class="headerlink" href="#module-genoml.nlp.kmers" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="genoml.nlp.kmers.Kmer">
<em class="property">class </em><code class="descclassname">genoml.nlp.kmers.</code><code class="descname">Kmer</code><a class="headerlink" href="#genoml.nlp.kmers.Kmer" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></p>
<p>Class to manager kmer objects</p>
<dl class="class">
<dt id="genoml.nlp.kmers.Kmer.Token">
<em class="property">class </em><code class="descname">Token</code><a class="headerlink" href="#genoml.nlp.kmers.Kmer.Token" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></p>
<p>Kmer subclass to manage kmer tokens</p>
<dl class="attribute">
<dt id="genoml.nlp.kmers.Kmer.Token.alphabet">
<code class="descname">alphabet</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.kmers.Kmer.Token.alphabet" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="genoml.nlp.kmers.Kmer.Token.complement">
<code class="descname">complement</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.kmers.Kmer.Token.complement" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.Kmer.Token.entropy">
<code class="descname">entropy</code><span class="sig-paren">(</span><em>sig=2</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.Kmer.Token.entropy" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute Shannon entropy for a given kmer corresponding to a token
:option sig (int): significant digits after decimal
:return entropy (float): Shannon entropy of kmer corresponding to a token</p>
</dd></dl>

<dl class="attribute">
<dt id="genoml.nlp.kmers.Kmer.Token.k">
<code class="descname">k</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.kmers.Kmer.Token.k" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="genoml.nlp.kmers.Kmer.Token.pos">
<code class="descname">pos</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.kmers.Kmer.Token.pos" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.Kmer.Token.to_kmer">
<code class="descname">to_kmer</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.Kmer.Token.to_kmer" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert token into its original kmer
:return token (string): token</p>
</dd></dl>

</dd></dl>

<dl class="attribute">
<dt id="genoml.nlp.kmers.Kmer.alphabet">
<code class="descname">alphabet</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.kmers.Kmer.alphabet" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="genoml.nlp.kmers.Kmer.complement">
<code class="descname">complement</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.kmers.Kmer.complement" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.Kmer.entropy">
<code class="descname">entropy</code><span class="sig-paren">(</span><em>sig=2</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.Kmer.entropy" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute Shannon entropy for a given kmer
:option sig (int): significant digits after decimal
:return entropy (float): Shannon entropy of kmer</p>
</dd></dl>

<dl class="attribute">
<dt id="genoml.nlp.kmers.Kmer.k">
<code class="descname">k</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.kmers.Kmer.k" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.Kmer.lower">
<code class="descname">lower</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.Kmer.lower" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert kmer to all lowercase
:return kmer_lower (Kmer): kmer in all lowercase</p>
</dd></dl>

<dl class="attribute">
<dt id="genoml.nlp.kmers.Kmer.pos">
<code class="descname">pos</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.kmers.Kmer.pos" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.Kmer.set_alphabet">
<code class="descname">set_alphabet</code><span class="sig-paren">(</span><em>alphabet</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.Kmer.set_alphabet" title="Permalink to this definition">¶</a></dt>
<dd><p>Set a new alphabet
:option alphabet (string): allowed alphabet for kmer</p>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.Kmer.set_complement">
<code class="descname">set_complement</code><span class="sig-paren">(</span><em>complement</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.Kmer.set_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Set a new complement alphabet
:option complement (string): complement to alphabet</p>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.Kmer.to_token">
<code class="descname">to_token</code><span class="sig-paren">(</span><em>complement=None</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.Kmer.to_token" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert kmer into token
:option complement (string): complement of Kmer alphabet
:return token (string): token corresponding to a kmer</p>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.Kmer.upper">
<code class="descname">upper</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.Kmer.upper" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert kmer to all uppercase
:return kmer_upper (Kmer): kmer in all uppercase</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="genoml.nlp.kmers.KmerList">
<em class="property">class </em><code class="descclassname">genoml.nlp.kmers.</code><code class="descname">KmerList</code><span class="sig-paren">(</span><em>arg</em>, <em>k=None</em>, <em>alphabet=None</em>, <em>complement=None</em>, <em>lower=True</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerList" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></p>
<p>Class to manage lists of kmers</p>
<dl class="class">
<dt id="genoml.nlp.kmers.KmerList.TokenList">
<em class="property">class </em><code class="descname">TokenList</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerList.TokenList" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></p>
<p>Class to manage lists of kmer tokens</p>
<dl class="method">
<dt id="genoml.nlp.kmers.KmerList.TokenList.all_tokens">
<code class="descname">all_tokens</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerList.TokenList.all_tokens" title="Permalink to this definition">¶</a></dt>
<dd><p>List all possible unique tokens
:return uniq_tokens (KmerList.TokenList): list of all sorted unique tokens</p>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerList.TokenList.append">
<code class="descname">append</code><span class="sig-paren">(</span><em>token</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerList.TokenList.append" title="Permalink to this definition">¶</a></dt>
<dd><p>Append new token to token list
:param token (Kmer.Token): token corresponding to a kmer</p>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerList.TokenList.entropy_stopwords">
<code class="descname">entropy_stopwords</code><span class="sig-paren">(</span><em>sig=2</em>, <em>entropy_threshold=1.3</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerList.TokenList.entropy_stopwords" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate set of stopwords according to Shannon entropy
:option entropy_threshold (float): Shannon entropy upper bound for stopwords
:return stopwords (KmerList.TokenList): list of sorted low-complexity tokens</p>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerList.TokenList.to_kmers">
<code class="descname">to_kmers</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerList.TokenList.to_kmers" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert tokens into their original kmers
:return kmers (string): kmers</p>
</dd></dl>

</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerList.all_kmers">
<code class="descname">all_kmers</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerList.all_kmers" title="Permalink to this definition">¶</a></dt>
<dd><p>List all possible unique kmers
:return uniq_kmers (KmerList): list of all sorted unique kmers</p>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerList.append">
<code class="descname">append</code><span class="sig-paren">(</span><em>kmer</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerList.append" title="Permalink to this definition">¶</a></dt>
<dd><p>Append new kmer to kmer list
:param kmer (Kmer): kmer</p>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerList.entropy_stopwords">
<code class="descname">entropy_stopwords</code><span class="sig-paren">(</span><em>sig=2</em>, <em>entropy_threshold=1.3</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerList.entropy_stopwords" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate set of stopwords according to Shannon entropy
:option entropy_threshold (float): Shannon entropy upper bound for stopwords
:return stopwords (KmerList): list of sorted low-complexity kmers</p>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerList.to_tokens">
<code class="descname">to_tokens</code><span class="sig-paren">(</span><em>complement=None</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerList.to_tokens" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert kmer list into token list
:option complement (string): complement of Kmer alphabet
:return _tokens (list): tokens corresponding to kmers in list</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-genoml.nlp">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-genoml.nlp" title="Permalink to this headline">¶</a></h2>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">genoml</a></h1>








<h3>Navigation</h3>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019, Ross Altman.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/source/genoml.nlp.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>