

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>genoml.nlp package &mdash; genoml 0.0.1 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> genoml
          

          
          </a>

          
            
            
              <div class="version">
                0.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <!-- Local TOC -->
              <div class="local-toc"><ul>
<li><a class="reference internal" href="#">genoml.nlp package</a><ul>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-genoml.nlp.bedtools">genoml.nlp.bedtools module</a></li>
<li><a class="reference internal" href="#module-genoml.nlp.kmers">genoml.nlp.kmers module</a></li>
<li><a class="reference internal" href="#module-genoml.nlp">Module contents</a></li>
</ul>
</li>
</ul>
</div>
            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">genoml</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>genoml.nlp package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/genoml.nlp.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="genoml-nlp-package">
<h1>genoml.nlp package<a class="headerlink" href="#genoml-nlp-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-genoml.nlp.bedtools">
<span id="genoml-nlp-bedtools-module"></span><h2>genoml.nlp.bedtools module<a class="headerlink" href="#module-genoml.nlp.bedtools" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="genoml.nlp.bedtools.BedTool">
<em class="property">class </em><code class="descclassname">genoml.nlp.bedtools.</code><code class="descname">BedTool</code><span class="sig-paren">(</span><em>fn=None</em>, <em>from_string=False</em>, <em>remote=False</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.bedtools.BedTool" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">pybedtools.bedtool.BedTool</span></code></p>
<p>Class to wrap pybedtools.BedTool for use with machine learning and
natural language processing</p>
<dl class="attribute">
<dt id="genoml.nlp.bedtools.BedTool.alphabet">
<code class="descname">alphabet</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.bedtools.BedTool.alphabet" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="genoml.nlp.bedtools.BedTool.complement">
<code class="descname">complement</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.bedtools.BedTool.complement" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="genoml.nlp.bedtools.BedTool.create_corpus">
<code class="descname">create_corpus</code><span class="sig-paren">(</span><em>intervals</em>, <em>k</em>, <em>fasta=None</em>, <em>upstream=0</em>, <em>downstream=0</em>, <em>offset=1</em>, <em>sep=''</em>, <em>processes=1</em>, <em>chunksize=1</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.bedtools.BedTool.create_corpus" title="Permalink to this definition">¶</a></dt>
<dd><p>Create corpus of kmer tokens from intervals</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>intervals</strong> (<em>list</em>) – list of interval indexes or pybedtools Interval objects</li>
<li><strong>k</strong> (<em>int</em>) – size of each kmer</li>
<li><strong>fasta</strong> (<em>str/io.TextIOWrapper</em>) – path to fasta file or stream (<em>default: None</em>)</li>
<li><strong>upstream</strong> (<em>int</em>) – number of bases upstream from each interval to read (<em>default: 0</em>)</li>
<li><strong>downstream</strong> (<em>int</em>) – number of bases downstream from each interval to read (<em>default: 0</em>)</li>
<li><strong>offset</strong> (<em>int</em>) – shift offset between kmers in each interval (<em>default: 1</em>)</li>
<li><strong>sep</strong> (<em>str</em>) – delimiter for token ‘words’</li>
<li><strong>processes</strong> (<em>int</em>) – number of worker processes to use</li>
<li><strong>chunksize</strong> (<em>int</em>) – size of each chunk submitted to a worker process</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">corpus of kmer tokens from intervals</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.bedtools.BedTool.read_expression_levels">
<code class="descname">read_expression_levels</code><span class="sig-paren">(</span><em>intervals</em>, <em>expr=None</em>, <em>sample_regex=None</em>, <em>processes=1</em>, <em>chunksize=1</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.bedtools.BedTool.read_expression_levels" title="Permalink to this definition">¶</a></dt>
<dd><p>Read expression levels for genes corresponding to all given intervals</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>intervals</strong> (<em>list</em>) – list of interval indexes or pybedtools Interval objects</li>
<li><strong>expr</strong> (<em>str</em>) – path to normalized feature count table (<em>default: None</em>)</li>
<li><strong>sample_regex</strong> (<em>str</em>) – regex to filter by sample name (<em>default: None</em>)</li>
<li><strong>processes</strong> (<em>int</em>) – number of worker processes to use</li>
<li><strong>chunksize</strong> (<em>int</em>) – size of each chunk submitted to a worker process</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">mean expression level for gene corresponding to interval</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">numpy.ndarray</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.bedtools.BedTool.read_kmers">
<code class="descname">read_kmers</code><span class="sig-paren">(</span><em>interval</em>, <em>k</em>, <em>fasta=None</em>, <em>upstream=0</em>, <em>downstream=0</em>, <em>offset=1</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.bedtools.BedTool.read_kmers" title="Permalink to this definition">¶</a></dt>
<dd><p>Read kmers from sequence from annotated interval in fasta file</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>interval</strong> (<em>int/Interval</em>) – index of interval or pybedtools Interval object</li>
<li><strong>k</strong> (<em>int</em>) – size of each kmer</li>
<li><strong>fasta</strong> (<em>str/io.TextIOWrapper</em>) – path to fasta file or stream (<em>default: None</em>)</li>
<li><strong>upstream</strong> (<em>int</em>) – number of bases upstream from interval to read (<em>default: 0</em>)</li>
<li><strong>downstream</strong> (<em>int</em>) – number of bases downstream from interval to read (<em>default: 0</em>)</li>
<li><strong>offset</strong> (<em>int</em>) – shift offset between kmers in interval (<em>default: 1</em>)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">list of kmers in interval</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="#genoml.nlp.kmers.KmerList" title="genoml.nlp.kmers.KmerList">KmerList</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.bedtools.BedTool.read_seq">
<code class="descname">read_seq</code><span class="sig-paren">(</span><em>interval</em>, <em>fasta=None</em>, <em>upstream=0</em>, <em>downstream=0</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.bedtools.BedTool.read_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Read sequence from annotated interval in fasta file</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>interval</strong> (<em>int/Interval</em>) – index of interval or pybedtools Interval object</li>
<li><strong>fasta</strong> (<em>str/io.TextIOWrapper</em>) – path to fasta file or stream (<em>default: None</em>)</li>
<li><strong>upstream</strong> (<em>int</em>) – number of bases upstream from interval to read (<em>default: 0</em>)</li>
<li><strong>downstream</strong> (<em>int</em>) – number of bases downstream from interval to read (<em>default: 0</em>)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">raw interval sequence from fasta file</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">string</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.bedtools.BedTool.read_tokens">
<code class="descname">read_tokens</code><span class="sig-paren">(</span><em>interval</em>, <em>k</em>, <em>fasta=None</em>, <em>upstream=0</em>, <em>downstream=0</em>, <em>offset=1</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.bedtools.BedTool.read_tokens" title="Permalink to this definition">¶</a></dt>
<dd><p>Read tokens from sequence from annotated interval in fasta file</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>interval</strong> (<em>int/Interval</em>) – index of interval or pybedtools Interval object</li>
<li><strong>k</strong> (<em>int</em>) – size of each kmer</li>
<li><strong>fasta</strong> (<em>str/io.TextIOWrapper</em>) – path to fasta file or stream (<em>default: None</em>)</li>
<li><strong>upstream</strong> (<em>int</em>) – number of bases upstream from interval to read (<em>default: 0</em>)</li>
<li><strong>downstream</strong> (<em>int</em>) – number of bases downstream from interval to read (<em>default: 0</em>)</li>
<li><strong>offset</strong> (<em>int</em>) – shift offset between kmers in interval (<em>default: 1</em>)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">list of kmer tokens in interval</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="#genoml.nlp.kmers.KmerList.TokenList" title="genoml.nlp.kmers.KmerList.TokenList">KmerList.TokenList</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.bedtools.BedTool.remove_invalid">
<code class="descname">remove_invalid</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.bedtools.BedTool.remove_invalid" title="Permalink to this definition">¶</a></dt>
<dd><p>Wrapper for pybedtools.BedTool.remove_invalid(*args, **kwargs)</p>
<p>Removes invalid entries from GTF/GFF files (i.e. chromosomes with
negative coordinates or features of zero length</p>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.bedtools.BedTool.saveas">
<code class="descname">saveas</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.bedtools.BedTool.saveas" title="Permalink to this definition">¶</a></dt>
<dd><p>Wrapper for pybedtools.BedTool.saveas(*args, **kwargs)</p>
<p>Creates tempfile to store intermediate data, allows IntervalIterator to reset</p>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.bedtools.BedTool.subset_feature">
<code class="descname">subset_feature</code><span class="sig-paren">(</span><em>feature_type</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.bedtools.BedTool.subset_feature" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract feature type from each gene</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>feature_type</strong> (<em>str</em>) – type of gene feature (e.g. five_prime_utr, three_prime_utr, exon, etc.)</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">pybedtools interval iterator limited to region_type intervals</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="#genoml.nlp.bedtools.BedTool" title="genoml.nlp.bedtools.BedTool">BedTool</a></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-genoml.nlp.kmers">
<span id="genoml-nlp-kmers-module"></span><h2>genoml.nlp.kmers module<a class="headerlink" href="#module-genoml.nlp.kmers" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="genoml.nlp.kmers.Kmer">
<em class="property">class </em><code class="descclassname">genoml.nlp.kmers.</code><code class="descname">Kmer</code><a class="headerlink" href="#genoml.nlp.kmers.Kmer" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></p>
<p>Class to manager kmer objects</p>
<dl class="class">
<dt id="genoml.nlp.kmers.Kmer.Token">
<em class="property">class </em><code class="descname">Token</code><a class="headerlink" href="#genoml.nlp.kmers.Kmer.Token" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></p>
<p>Kmer subclass to manage kmer tokens</p>
<dl class="attribute">
<dt id="genoml.nlp.kmers.Kmer.Token.alphabet">
<code class="descname">alphabet</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.kmers.Kmer.Token.alphabet" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="genoml.nlp.kmers.Kmer.Token.complement">
<code class="descname">complement</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.kmers.Kmer.Token.complement" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.Kmer.Token.entropy">
<code class="descname">entropy</code><span class="sig-paren">(</span><em>sig=2</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.Kmer.Token.entropy" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute Shannon entropy for a given kmer corresponding to a token</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>sig</strong> (<em>int</em>) – significant digits after decimal (<em>default: 2</em>)</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Shannon entropy of kmer corresponding to a token</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">float</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="genoml.nlp.kmers.Kmer.Token.k">
<code class="descname">k</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.kmers.Kmer.Token.k" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="genoml.nlp.kmers.Kmer.Token.pos">
<code class="descname">pos</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.kmers.Kmer.Token.pos" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.Kmer.Token.to_kmer">
<code class="descname">to_kmer</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.Kmer.Token.to_kmer" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert token into its original kmer</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">token</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">string</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="attribute">
<dt id="genoml.nlp.kmers.Kmer.alphabet">
<code class="descname">alphabet</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.kmers.Kmer.alphabet" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="genoml.nlp.kmers.Kmer.complement">
<code class="descname">complement</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.kmers.Kmer.complement" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.Kmer.entropy">
<code class="descname">entropy</code><span class="sig-paren">(</span><em>sig=2</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.Kmer.entropy" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute Shannon entropy for a given kmer</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>sig</strong> (<em>int</em>) – significant digits after decimal (<em>default: 2</em>)</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Shannon entropy of kmer</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">float</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="genoml.nlp.kmers.Kmer.k">
<code class="descname">k</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.kmers.Kmer.k" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.Kmer.lower">
<code class="descname">lower</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.Kmer.lower" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert kmer to all lowercase</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">kmer in all lowercase</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="#genoml.nlp.kmers.Kmer" title="genoml.nlp.kmers.Kmer">Kmer</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="genoml.nlp.kmers.Kmer.pos">
<code class="descname">pos</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.kmers.Kmer.pos" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="genoml.nlp.kmers.Kmer.sep">
<code class="descname">sep</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.kmers.Kmer.sep" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.Kmer.set_alphabet">
<code class="descname">set_alphabet</code><span class="sig-paren">(</span><em>alphabet</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.Kmer.set_alphabet" title="Permalink to this definition">¶</a></dt>
<dd><p>Set a new alphabet</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>alphabet</strong> (<em>str</em>) – allowed alphabet for kmer</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.Kmer.set_complement">
<code class="descname">set_complement</code><span class="sig-paren">(</span><em>complement</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.Kmer.set_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Set a new complement alphabet</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>complement</strong> (<em>str</em>) – complement to alphabet</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.Kmer.to_token">
<code class="descname">to_token</code><span class="sig-paren">(</span><em>complement=None</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.Kmer.to_token" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert kmer into token</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>complement</strong> (<em>str</em>) – complement of Kmer alphabet (<em>default: None</em>)</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">token corresponding to a kmer</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">string</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.Kmer.upper">
<code class="descname">upper</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.Kmer.upper" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert kmer to all uppercase</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">kmer in all uppercase</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="#genoml.nlp.kmers.Kmer" title="genoml.nlp.kmers.Kmer">Kmer</a></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="genoml.nlp.kmers.KmerCorpus">
<em class="property">class </em><code class="descclassname">genoml.nlp.kmers.</code><code class="descname">KmerCorpus</code><span class="sig-paren">(</span><em>arg</em>, <em>k=None</em>, <em>alphabet=None</em>, <em>complement=None</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerCorpus" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></p>
<p>Class to manage corpora of kmer tokens</p>
<dl class="class">
<dt id="genoml.nlp.kmers.KmerCorpus.TokenSentence">
<em class="property">class </em><code class="descname">TokenSentence</code><a class="headerlink" href="#genoml.nlp.kmers.KmerCorpus.TokenSentence" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></p>
<p>Class to manage lists of kmer tokens</p>
<dl class="method">
<dt id="genoml.nlp.kmers.KmerCorpus.TokenSentence.all_tokens">
<code class="descname">all_tokens</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerCorpus.TokenSentence.all_tokens" title="Permalink to this definition">¶</a></dt>
<dd><p>List all possible unique tokens</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list of all sorted unique tokens</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="#genoml.nlp.kmers.KmerList.TokenList" title="genoml.nlp.kmers.KmerList.TokenList">KmerList.TokenList</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="genoml.nlp.kmers.KmerCorpus.TokenSentence.alphabet">
<code class="descname">alphabet</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.kmers.KmerCorpus.TokenSentence.alphabet" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="genoml.nlp.kmers.KmerCorpus.TokenSentence.complement">
<code class="descname">complement</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.kmers.KmerCorpus.TokenSentence.complement" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerCorpus.TokenSentence.entropy_stopwords">
<code class="descname">entropy_stopwords</code><span class="sig-paren">(</span><em>sig=2</em>, <em>entropy_threshold=1.3</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerCorpus.TokenSentence.entropy_stopwords" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate set of stopwords according to Shannon entropy</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>sig</strong> (<em>int</em>) – Shannon entropy upper bound for stopwords (<em>default: 2</em>)</li>
<li><strong>entropy_threshold</strong> (<em>float</em>) – Shannon entropy upper bound for stopwords (<em>default: 1.3</em>)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">list of sorted low-complexity tokens</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="#genoml.nlp.kmers.KmerList.TokenList" title="genoml.nlp.kmers.KmerList.TokenList">KmerList.TokenList</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="genoml.nlp.kmers.KmerCorpus.TokenSentence.k">
<code class="descname">k</code><em class="property"> = None</em><a class="headerlink" href="#genoml.nlp.kmers.KmerCorpus.TokenSentence.k" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerCorpus.TokenSentence.to_kmers">
<code class="descname">to_kmers</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerCorpus.TokenSentence.to_kmers" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert tokens into their original kmers</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">kmers</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="#genoml.nlp.kmers.KmerList" title="genoml.nlp.kmers.KmerList">KmerList</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerCorpus.TokenSentence.to_tokens">
<code class="descname">to_tokens</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerCorpus.TokenSentence.to_tokens" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert kmers to tokens</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">tokens</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="#genoml.nlp.kmers.KmerList.TokenList" title="genoml.nlp.kmers.KmerList.TokenList">TokenList</a></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerCorpus.all_kmers">
<code class="descname">all_kmers</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerCorpus.all_kmers" title="Permalink to this definition">¶</a></dt>
<dd><p>List all possible unique kmers</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list of all sorted unique kmers</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="#genoml.nlp.kmers.KmerList" title="genoml.nlp.kmers.KmerList">KmerList</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerCorpus.all_tokens">
<code class="descname">all_tokens</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerCorpus.all_tokens" title="Permalink to this definition">¶</a></dt>
<dd><p>List all possible unique tokens</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list of all sorted unique tokens</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="#genoml.nlp.kmers.KmerList.TokenList" title="genoml.nlp.kmers.KmerList.TokenList">KmerList.TokenList</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerCorpus.append">
<code class="descname">append</code><span class="sig-paren">(</span><em>token_sentence</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerCorpus.append" title="Permalink to this definition">¶</a></dt>
<dd><p>Append new KmerCorpus.TokenSentence to KmerCorpus</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>token_sentence</strong> (<a class="reference internal" href="#genoml.nlp.kmers.KmerCorpus.TokenSentence" title="genoml.nlp.kmers.KmerCorpus.TokenSentence"><em>KmerCorpus.TokenSentence</em></a>) – kmer token sentence</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerCorpus.entropy_stopwords">
<code class="descname">entropy_stopwords</code><span class="sig-paren">(</span><em>sig=2</em>, <em>entropy_threshold=1.3</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerCorpus.entropy_stopwords" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate set of stopwords according to Shannon entropy</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>sig</strong> (<em>int</em>) – Shannon entropy upper bound for stopwords (<em>default: 2</em>)</li>
<li><strong>entropy_threshold</strong> (<em>float</em>) – Shannon entropy upper bound for stopwords (<em>default: 1.3</em>)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">list of sorted low-complexity tokens</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="#genoml.nlp.kmers.KmerList.TokenList" title="genoml.nlp.kmers.KmerList.TokenList">KmerList.TokenList</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerCorpus.extend">
<code class="descname">extend</code><span class="sig-paren">(</span><em>kmer_corpus</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerCorpus.extend" title="Permalink to this definition">¶</a></dt>
<dd><p>Extend KmerCorpus with another KmerCorpus</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>kmer_corpus</strong> (<a class="reference internal" href="#genoml.nlp.kmers.KmerCorpus" title="genoml.nlp.kmers.KmerCorpus"><em>KmerCorpus</em></a>) – kmer token corpus</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="genoml.nlp.kmers.KmerList">
<em class="property">class </em><code class="descclassname">genoml.nlp.kmers.</code><code class="descname">KmerList</code><span class="sig-paren">(</span><em>arg</em>, <em>k=None</em>, <em>alphabet=None</em>, <em>complement=None</em>, <em>lower=True</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerList" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></p>
<p>Class to manage lists of kmers</p>
<dl class="class">
<dt id="genoml.nlp.kmers.KmerList.TokenList">
<em class="property">class </em><code class="descname">TokenList</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerList.TokenList" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></p>
<p>Class to manage lists of kmer tokens</p>
<dl class="method">
<dt id="genoml.nlp.kmers.KmerList.TokenList.all_tokens">
<code class="descname">all_tokens</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerList.TokenList.all_tokens" title="Permalink to this definition">¶</a></dt>
<dd><p>List all possible unique tokens</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list of all sorted unique tokens</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="#genoml.nlp.kmers.KmerList.TokenList" title="genoml.nlp.kmers.KmerList.TokenList">KmerList.TokenList</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerList.TokenList.append">
<code class="descname">append</code><span class="sig-paren">(</span><em>token</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerList.TokenList.append" title="Permalink to this definition">¶</a></dt>
<dd><p>Append new Kmer.Token to KmerList.TokenList</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>token</strong> (<a class="reference internal" href="#genoml.nlp.kmers.Kmer.Token" title="genoml.nlp.kmers.Kmer.Token"><em>Kmer.Token</em></a>) – token corresponding to a kmer</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerList.TokenList.entropy_stopwords">
<code class="descname">entropy_stopwords</code><span class="sig-paren">(</span><em>sig=2</em>, <em>entropy_threshold=1.3</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerList.TokenList.entropy_stopwords" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate set of stopwords according to Shannon entropy</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>sig</strong> (<em>int</em>) – Shannon entropy upper bound for stopwords (<em>default: 2</em>)</li>
<li><strong>entropy_threshold</strong> (<em>float</em>) – Shannon entropy upper bound for stopwords (<em>default: 1.3</em>)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">list of sorted low-complexity tokens</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="#genoml.nlp.kmers.KmerList.TokenList" title="genoml.nlp.kmers.KmerList.TokenList">KmerList.TokenList</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerList.TokenList.extend">
<code class="descname">extend</code><span class="sig-paren">(</span><em>tokens</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerList.TokenList.extend" title="Permalink to this definition">¶</a></dt>
<dd><p>Extend KmerList.TokenList with another KmerList.TokenList</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>tokens</strong> (<a class="reference internal" href="#genoml.nlp.kmers.KmerList.TokenList" title="genoml.nlp.kmers.KmerList.TokenList"><em>KmerList.TokenList</em></a>) – tokens corresponding to kmers</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerList.TokenList.to_kmers">
<code class="descname">to_kmers</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerList.TokenList.to_kmers" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert tokens into their original kmers</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">kmers</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="#genoml.nlp.kmers.KmerList" title="genoml.nlp.kmers.KmerList">KmerList</a></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerList.all_kmers">
<code class="descname">all_kmers</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerList.all_kmers" title="Permalink to this definition">¶</a></dt>
<dd><p>List all possible unique kmers</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list of all sorted unique kmers</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="#genoml.nlp.kmers.KmerList" title="genoml.nlp.kmers.KmerList">KmerList</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerList.append">
<code class="descname">append</code><span class="sig-paren">(</span><em>kmer</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerList.append" title="Permalink to this definition">¶</a></dt>
<dd><p>Append new Kmer to KmerList</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>kmer</strong> (<a class="reference internal" href="#genoml.nlp.kmers.Kmer" title="genoml.nlp.kmers.Kmer"><em>Kmer</em></a>) – kmer</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerList.entropy_stopwords">
<code class="descname">entropy_stopwords</code><span class="sig-paren">(</span><em>sig=2</em>, <em>entropy_threshold=1.3</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerList.entropy_stopwords" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate set of stopwords according to Shannon entropy</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>sig</strong> (<em>int</em>) – Shannon entropy upper bound for stopwords (<em>default: 2</em>)</li>
<li><strong>entropy_threshold</strong> (<em>float</em>) – Shannon entropy upper bound for stopwords (<em>default: 1.3</em>)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">list of sorted low-complexity tokens</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="#genoml.nlp.kmers.KmerList.TokenList" title="genoml.nlp.kmers.KmerList.TokenList">KmerList.TokenList</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerList.extend">
<code class="descname">extend</code><span class="sig-paren">(</span><em>kmers</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerList.extend" title="Permalink to this definition">¶</a></dt>
<dd><p>Extend KmerList with another KmerList</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>kmer</strong> (<a class="reference internal" href="#genoml.nlp.kmers.Kmer" title="genoml.nlp.kmers.Kmer"><em>Kmer</em></a>) – kmer</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="genoml.nlp.kmers.KmerList.to_tokens">
<code class="descname">to_tokens</code><span class="sig-paren">(</span><em>complement=None</em><span class="sig-paren">)</span><a class="headerlink" href="#genoml.nlp.kmers.KmerList.to_tokens" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert kmer list into token list</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>complement</strong> (<em>str</em>) – complement of Kmer alphabet (<em>default: None</em>)</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">tokens corresponding to kmers in list</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="#genoml.nlp.kmers.KmerList.TokenList" title="genoml.nlp.kmers.KmerList.TokenList">KmerList.TokenList</a></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-genoml.nlp">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-genoml.nlp" title="Permalink to this headline">¶</a></h2>
</div>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Ross Altman

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
    

  

  <script type="text/javascript" src="_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>